packages/SwingSet/bin/extract-transcript-from-kerneldb.js

```diff
@@ -0,0 +1,109 @@
+import lmdb from 'node-lmdb';
+import process from 'process';
+import fs from 'fs';
+
+const argv = process.argv.splice(2);
+const dirPath = argv[0];
+const vatName = argv[1];
+
+if (!dirPath) {
+  console.log('extract-transcript-from-kerneldb DBDIR : list all vats in DB');
+  console.log(
+    'extract-transcript-from-kerneldb DBDIR VATID : extract transcript',
+  );
+  process.exit(0);
+}
+
+const lmdbEnv = new lmdb.Env();
+lmdbEnv.open({
+  path: dirPath,
+  mapSize: 2 * 1024 * 1024 * 1024, // XXX need to tune this
+});
+
+const dbi = lmdbEnv.openDbi({
+  name: 'swingset-kernel-state',
+  create: false,
+});
+const txn = lmdbEnv.beginTxn();
+function get(key) {
+  return txn.getString(dbi, key);
+}
+
+const allVatNames = JSON.parse(get('vat.names'));
+const allDynamicVatIDs = JSON.parse(get('vat.dynamicIDs'));
+
+if (!vatName) {
+  console.log(`all vats:`);
+  for (const name of allVatNames) {
+    const vatID = get(`vat.name.${name}`);
+    const transcriptLength = Number(get(`${vatID}.t.nextID`));
+    console.log(`${vatID} : ${name}       (${transcriptLength} deliveries)`);
+  }
+  for (const vatID of allDynamicVatIDs) {
+    const transcriptLength = Number(get(`${vatID}.t.nextID`));
+    console.log(
+      `${vatID} : (dynamic)`,
+      get(`${vatID}.options`),
+      `   (${transcriptLength} deliveries)`,
+    );
+  }
+} else {
+  let vatID = vatName;
+  if (allVatNames.indexOf(vatName) !== -1) {
+    vatID = get(`vat.name.${vatName}`);
+  }
+  if (!get(`${vatID}.options`)) {
+    throw Error(`unable to find vatID ${vatID}`);
+  }
+  console.log();
+  const fn = `transcript-${vatID}.sst`;
+  console.log(`extracting transcript for vat ${vatID} into ${fn}`);
+  const fd = fs.openSync(fn, 'w');
+
+  const dynamic = allDynamicVatIDs.includes(vatID);
+  const source = JSON.parse(get(`${vatID}.source`));
+  const vatSourceBundle = source.bundle || get(`bundle.${source.bundleName}`);
+  const options = JSON.parse(get(`${vatID}.options`));
+  console.log(`options:`, options);
+  const { vatParameters } = options;
+  console.log(`vatParameters:`, vatParameters);
+  let transcriptNum = 0;
+  const first = {
+    type: 'create-vat',
+    transcriptNum,
+    vatID,
+    dynamic,
+    vatParameters,
+    vatSourceBundle,
+  };
+  transcriptNum += 1;
+  // first line of transcript is the source bundle
+  fs.writeSync(fd, JSON.stringify(first));
+  fs.writeSync(fd, '\n');
```

Comment:
I'm wondering if the migration of transcripts out of the kernel DB key-value store should also do this.  In that case, it might render this entire tool unnecessary (or, rather, reduce to copying a file).

Reply:
(for the future:) we talked this through, and decided:

* the `.sst` transcript files should be standalone: you should be able to replay one without having access to any of the original database or slogfile or source files
* so the .sst file must contain all the vat sources you'll need, in the first couple of lines
* right now, sources are recorded as a big string, but in the future we'll be splitting them up into a package graph, probably stored as one-file-per-package, named after the hash of the package contents
  * the swingset DB will grow to track these package blobs in a side directory, just like we're about to do with the transcripts
  * when that happens, we'll change the `v$NN.source` entry in the key-value store, from using a large source bundle as the value, to using a small package ID string as the value
  * we'll probably also move the kernel source bundle and the various xsnap supervisor bundles out of the key-value store and into a blobstore, using similar IDs
* at that point, the transcript-extraction tool will need to copy all the necessary package blobs into the `.sst` file, before populating it with deliveries

So, we plan to leave this format alone for now. The upcoming transcript stream portion of the kernelDB will store only deliveries. The extraction tool will copy both the key-value `v$NN.source` bundle and the transcript deliveries into the `.sst` file.